### PR TITLE
[Gekidou] [MM-39936] Add Select Team Screen

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Add Select Team Screen

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-39936

#### Device Information
This PR was tested on:
- Android Emulator

#### Screenshots
Screenshot from 2022-04-20 12-00-15
Screenshot from 2022-04-20 12-01-57
Screenshot from 2022-04-20 12-02-40
Screenshot from 2022-04-20 12-02-59
Screenshot from 2022-04-20 12-21-15

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by larkox** [comment]
Pending iOS and Tablet testing / fixes.

**Comment by larkox** [comment]
Ready to review, just missing using the right illustration.

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Nice work @larkox. It's looking great - just a few things I noticed in my review:

1. After viewing the 'Select a team' screen, I switched to a different server (community server) and it showed me the 'Select a team' screen for that server for some reason. I the community server already setup previously in the app so it s should have taken mi directly to the homescreen rather than the Select a team screen.

2. I changed my test server permissions to prevent non-admin users from being able to create a team, but the 'Create a New Team' button still showed for non-admin users on the "No teams are available to join" screen.

3. On the "No teams are available to join" screen, the text in the 'Create a new team' button is not vertically aligned with the `+` icon and is off-center.
image

4. Similar to 2 above, the 'Create a New Team' button showed on the 'Select a team' screen, but shouldn't have since I disabled the new team permission. 

5. Can we add space of `8` above _and_ below the divider that shows under 'Create a New Team' on the 'Select a team' screen?

6. Can we change the label on the 'Create a New Team' button to be sentence case? `Create a new team` (lowercase `n` and `t`) - this applies to both the 'Select a team' screen and the 'Not teams are available to join' screen.

7. On the 'Select a team' screen, can we reduce the margin on the left and right by 8?


8. For the Server button in the top left, can we make the label tappable as well as the icon? Right now it seems that the icon is only tappable.

9. As discussed, the illustration needs to be updated. We can chat about this over DM though

10. I assume we need to add websocket events for team changes? When I made changes to team permissions on the system console, those changes weren't reflected for the end user until they switched to a different server and returned

**Comment by larkox** [comment]
Thanks for the thorough review! I will address all your points. It seems there is some transitions that I have not handled well yet (like switching between servers). Also, feels like something is weird with the permissions (not sure if websocket related, or observables related). A "hard refresh" (closing the app and opening) should solve the permission issues (I hope).

The server touchable I have to look a bit deeper. The component is shared in other places, so I have to make sure the changes "make sense" everywhere.

For the cosmetic changes, they should be no problem.

File: app/screens/select_team/select_team.tsx
Comment (avinashlng1080) [inline app/screens/select_team/select_team.tsx:96]:
```suggestion
    
    useEffect(() => {
```

File: app/screens/select_team/select_team.tsx
Comment (avinashlng1080) [inline app/screens/select_team/select_team.tsx:123]:
```suggestion
    
    return (
```

File: app/screens/select_team/select_team.tsx
Comment (avinashlng1080) [inline app/screens/select_team/select_team.tsx:71]:
```suggestion
```

File: app/screens/select_team/no_teams.tsx
Comment (avinashlng1080) [inline app/screens/select_team/no_teams.tsx:37]:
```suggestion
```

File: app/screens/select_team/add_team_item.tsx
Comment (avinashlng1080) [inline app/screens/select_team/add_team_item.tsx:57]: Will this be implemented in another ticket and can you please add the ticket number also?

File: app/components/team_sidebar/team_list/team_item/team_icon.tsx
Comment (avinashlng1080) [reply app/components/team_sidebar/team_list/team_item/team_icon.tsx:1]: The style container can be deleted.

File: app/screens/home/channel_list/channel_list.tsx
Comment (avinashlng1080) [reply app/screens/home/channel_list/channel_list.tsx:72]: the style flex can be removed.

**Review by avinashlng1080 — CHANGES_REQUESTED** [review]
Nice PR with some minor updates.

File: app/components/team_sidebar/team_list/team_item/team_icon.tsx
Comment (larkox) [reply app/components/team_sidebar/team_list/team_item/team_icon.tsx:1]: Both container and containerSelected are used. Right now I think it is simpler to leave the as they are instead of trying to merge them into a single one.

**Comment by larkox** [comment]
@matthewbirtch
Regarding 2 and 4:
This seems to be working on my side. Do you mind sharing more in depth steps? (In case I am missing something)
Regardin 7:
Just solved. CSS was being unruly here :sweat_smile:. If I set 16 margin, one more word fitted the space, and made the space the "expected width". But if I set the margin to 24, that word had to go in the next line, and suddenly the width just hugged the text, and since it was centered (specially for tablets), it got "lots of margin" instead of just 24. Now should be fine.

Regarding 10:
With this you mean the errors from 2 and 4? Or you mean something different?

**Comment by matthewbirtch** [comment]
> Regarding 2 and 4: This seems to be working on my side. Do you mind sharing more in depth steps? (In case I am missing something)

Here's what I did:

- Admin user: go to `System console > Teams` and remove the end user from all teams, then save.
- Admin user: go to `System Console > Permissions > Edit Scheme` and under `All Members` uncheck `Create Teams`, then save.
- End user: Open RNv2 app and switch to server that changes occurred in. I get the 'No teams available' screen, but it shows the 'Create a new team' button

**Comment by larkox** [comment]
Regarding 2 and 4, the issue was that the roles were not updating correctly on startup. I saw it working right with the websockets, but didn't think about testing the "entry route". Should be fixed now.

Regarding 10, I also fixed a few small things. The only thing that should not update "instantly" is if you are in the select teams screen, and a new public (or accessible) team is created (it won't show in the list until you "refresh" (close the app, or go and back from other server)) or deleted (it won't disappear from the list). This is because there is no "a new accessible team has been created in the system" websocket event. If you get added to a team by any mean (added by other user, join in a different client...), it should go without warning into the team you just were added to.

**Comment by matthewbirtch** [comment]
> Regarding 2 and 4, the issue was that the roles were not updating correctly on startup. I saw it working right with the websockets, but didn't think about testing the "entry route". Should be fixed now.

Yep, the above is fixed now.

> Regarding 10, I also fixed a few small things. The only thing that should not update "instantly" is if you are in the select teams screen, and a new public (or accessible) team is created (it won't show in the list until you "refresh" (close the app, or go and back from other server)) or deleted (it won't disappear from the list). This is because there is no "a new accessible team has been created in the system" websocket event. If you get added to a team by any mean (added by other user, join in a different client...), it should go without warning into the team you just were added to.

I'm still not seeing instantaneous updates on anything to do with teams. Can we create a separate ticket to look in to this? I think we can merge this as-is and then look at the websocket issues separately. I don't want to block this merge.

**Review by matthewbirtch — APPROVED** [review]
Approving this assuming we open a separate ticket for the websocket stuff as mentioned above

**Review by avinashlng1080 — APPROVED** [review]
LGTM.  Just one minor nit

File: app/screens/select_team/no_teams.tsx
Comment (larkox) [reply app/screens/select_team/no_teams.tsx:38]: It is used in line 83.

File: app/store/ephemeral_store.ts
Comment (enahum) [inline app/store/ephemeral_store.ts:17]: you should set a getter and a removal function to access this prop... the way is currently done can be error prone, thus the reason why have getters and setters for the other properties as well.

File: app/actions/remote/team.ts
Comment (enahum) [inline app/actions/remote/team.ts:41]: see comment on `ephemeral_store`

File: app/actions/websocket/channel.ts
Comment (enahum) [inline app/actions/websocket/channel.ts:254]: see comment on `ephemeral_store`

File: app/actions/websocket/teams.ts
Comment (enahum) [inline app/actions/websocket/teams.ts:91]: same

File: app/queries/servers/role.ts
Comment (enahum) [inline app/queries/servers/role.ts:56]: should we change this to `observeMyTeam` instead of the `catchError` ?

File: app/screens/select_team/add_team_item.tsx
Comment (enahum) [inline app/screens/select_team/add_team_item.tsx:58]: Perhaps uncomment this line so that there is an alert being displayed when users tap on the button?

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/queries/servers/role.ts
Comment (enahum) [inline app/queries/servers/role.ts:73]: When are we going to hit the error here? this stops all observability as defaultValue won't get updated ever, right?

**Review by enahum — APPROVED** [review]

File: app/queries/servers/role.ts
Comment (larkox) [reply app/queries/servers/role.ts:73]: Missed it. Removing it now.

**Pull request merged**
